### PR TITLE
Distributed backend for Alibi

### PR DESCRIPTION
**Pull request by alexcoca**
This pull request builds on the ideas introduced in #270 in order to build a distributed backend for `alibi` using the `ray` library. 

**New objects**

1. `DistributedExplainer` class in `alibi.utils.distributed`. The role of this class is to create a pool of explainer replicas and distribute a batch of instances to be explained across the pool.

2. `PoolCollection` class in `alibi.utils.distributed`. This class is used when a particular explainer involves solving multiple independent optimisation problems. It instantiates `DistributedExplainer` objects in different processes, which allows us to distributed batches across different optimisation processes, simultaneously.

**Things to note**

1. The `DistributedExplainer` object has a `return_generator` option, which allows the user to return a generator which yields the computed explanations as they are returned. In this way, the caller can use the results of the computations before the whole batch is completed.

2. The above cannot be used in conjunction with the `PoolCollection` class, since this would involve pickling a generator (not supported in CPython). This suggests that the mechanism for returning results as they complete could be achieved by passing an appropriate callback to the `DistributedExplainer` object directly. That is, the generator would be consumed inside the `DistributedExplainer` by a function that can return the results to a given context (e.g., an endpoint).

### CODE REVIEW

**Comment by jklaise** [comment]
Should we use Ray's implementation of `ActorPool` going forward?

File: alibi/utils/distributed.py
Comment (jklaise) [inline alibi/utils/distributed.py:523]: I'm not sure I follow, how is `PoolCollection` related to this?

File: alibi/utils/tests/test_distributed.py
Comment (jklaise) [inline alibi/utils/tests/test_distributed.py:260]: Test the...?

File: alibi/utils/tests/test_distributed.py
Comment (jklaise) [inline alibi/utils/tests/test_distributed.py:322]: Test the...?

File: alibi/utils/tests/test_distributed.py
Comment (jklaise) [inline alibi/utils/tests/test_distributed.py:357]: setting what?

**Review by jklaise — CHANGES_REQUESTED** [review]
Nice work Alex! Let's set some time aside to go through the intended use cases as per one of my comments.

**Comment by alexcoca** [comment]
> Should we use Ray's implementation of `ActorPool` going forward?

The short answer is yes. I took the `alibi.utils.distributed.py` version from a branch in the `ray` repo before it was merged and added batching capability to it in order to implemenent the distributed anchor tabular algorithm.

Things to note:

- Batching logic will have to be implemented outside the `ActorPool`  object if you decide to only use `ray.util.ActorPool` simply importing from `ray.util` as opposed to our code will break the achors code

**Comment by alexcoca** [comment]
> Should we use Ray's implementation of `ActorPool` going forward?

File: alibi/utils/distributed.py
Comment (alexcoca) [reply alibi/utils/distributed.py:438]: Differences:

- target function: used by ray to schedule a remote explainer call (see `submit` method in `ray.util.ActorPool`

- postprocessing function: used to re-assemble the results computed in parallel into a single object (think e.g., concatenation of numpy arrays in the simplest case). It needs to be custom because each explainer returns something different. We can have a generic one that assumes the input array is a list numpy arrays and concatenates them, and register different functions for algorithms with more complex output types (e.g., kernel shap). 

[NB: On second thoughts I am thinking that `singledispatch` might be another way to implement this]

File: alibi/utils/tests/test_distributed.py
Comment (alexcoca) [reply alibi/utils/tests/test_distributed.py:139]: Sorry, this is an errant TODO. When I first wrote the test I was calling `ray.get` to obtain the attribute values for the assertion but subsequently realised that we should return the actual attributes, not futures.

File: alibi/utils/tests/test_distributed.py
Comment (alexcoca) [reply alibi/utils/tests/test_distributed.py:146]: When you want to completely tear down ray (e.g., kill the underlying processes, terminate actors, etc). This should be done in between tests to ensure corectness. Also worth checking out the ray docs h[ere](https://docs.ray.io/en/master/package-ref.html)

File: alibi/utils/distributed.py
Comment (alexcoca) [reply alibi/utils/distributed.py:541]: In brief:

- The underlying explainer doesn't need to work with minibatches if this does not make sense - simply set the `batch_size` to `1` in the default distributed options for the algorithms that don't do minibatches and raise an error if user overrides it. However, if in the future we have batched implementations that leverage vectorisation, we don't need to change the code at all.

- we have a fixed resource pool specifed by the user in the `distributed_opts` kwarg. If we have many more instances to explain, we queue them and feed them to the pool as some of the actors there finish the job. In this way we keep the number of processes under control

- if we want to play around with different `kwargs` for `explain` (aka, esentially run experiments in parallel), the most appropriate way would be to use the `PoolCollection` object - suggest a simple config file approach where the user can specify the algorithm and its configuration in a yaml and then run distributed experiments. Foundation work has been done in [this ](https://github.com/alexcoca/alibi/blob/Benchmarking/benchmark) branch.

File: alibi/utils/distributed.py
Comment (alexcoca) [reply alibi/utils/distributed.py:523]: Sorry, this is garbled. You can do `my_pool_collection.prop` where `prop` is a either an attribute of the `PoolCollection` or of the explainer wrapped by the `DistributedExplainer`. `__geattr__` handles the latter case.

**Review by jklaise — APPROVED** [review]
Looks great, happy to merge when tests pass.

**Pull request merged**
